RELEASES.md

```diff
@@ -0,0 +1,414 @@
+            to: Some(env.caller()),
+            value: init_value
+        });
+    }
+}
+```
+
+</td>
+<td>
+
+```rust
+impl Erc20 {
+    #[ink(constructor)]
+    fn new(&mut self, initial_supply: Balance) {
+        let caller = self.env().caller();
+        self.total_supply.set(initial_supply);
+        self.balances.insert(caller, initial_supply);
+        self.env().emit_event(Transferred {
+            from: None,
+            to: Some(caller),
+            amount: initial_supply,
+        });
+    }
+}
+```
+
+</td>
+</tr>
+</table>
+
+## Cross Contract Calls
+
+It is now possible to call ink! messages and ink! constructors. So ink! constructors allow
+delegation and ink! messages can easily call other ink! messages.
+
+Given another ink! contract like `mod Accumulator { ... }`, we can call any of its functions just
+like we would expect:
+
+```rust
+use accumulator::Accumulator;
+let result = Accumulator::my_function()
+```
+
+## Factory Contracts
+
+Creation of other contracts from a factory contract works pretty much the same way it did in the old
+ink! language.
+
+However, users are now required to specify the `code_hash` separately rather than in the
+constructor:
+
+```rust
+.using_code(code_hash)
+```
+
+Also, they need to specify the used ink! environment (most likely `self.env()`):
+
+```rust
+create_using(self.env())
+```
+
+<table style="width: 100%;">
+<tr>
+<th>Before</th>
+<th>After</th>
+</tr>
+<tr>
+<td>
+
+```rust
+let accumulator = Accumulator::new(accumulator_code_hash, init_value)
+    .value(total_balance / 4)
+    .create()
+    .expect("failed at instantiating the accumulator contract");
+```
+
+</td>
+<td>
+
+```rust
+let accumulator = Accumulator::new(init_value)
+    .value(total_balance / 4)
+    .using_code(accumulator_code_hash)
+    .create_using(self.env())
+    .expect("failed at instantiating the `Accumulator` contract");
+```
+
+</td>
+</tr>
+</table>
+
+## Contract Tests
+
+Testing contracts off-chain is done by `cargo test` and users can simply use the standard routines
+of creating unit test modules within the ink! module itself:
+
+```rust
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn my_test() { ... }
+}
+```
+
+Test instances of contracts can be created with something like:
+
+```rust
+let contract = MyContract::my_constructor(a, b);
+```
+
+Messages can simply be called on the returned instance as if `MyContract::my_constructor` returns a
+`Self` instance.
+
+**The off-chain test environment has lost a bit of power compared to the old ink! language.**
+
+It is not currently possible to query and set special test data about the environment (such as the
+caller of a function or amount of value sent), but these will be added back in the near future.
+
+## ink!-less Implementations
+
+It is also possible to annotate an entire impl blocks with:
+
+```rust
+#[ink(impl)]
+impl Contract {
+    fn internal_function(&self) {
+        self.env().emit_event(Hi);
```

Comment:
It might be confusing to have `self.env().emit_event(Hi);` here

Reply:
It is technically not wrong though right? There could be a defined event `Hi` elsewhere?

Reply:
Yeah but I think this `Hi` is rather confusing to beginners.

Reply:
Updated to `EventName`